Create documentation for connecting Astro to external data services (#480)
astronomer/docs: astro/connectivity.md

---
sidebar_label: "Connect to External Services"
title: "Connect Astro to External Data Services"
id: connectivity
description: Learn how to connect Astro to different types of external data sources.
---

## Overview

Before you can orchestrate production-level data in Astro, you first need to make this data accessible to your Astro Clusters. This guide explains how to connect your Astro Clusters to external data sources using the following methods:

- Public Endpoints
- VPC Peering

## Public Endpoints

The fastest and easiest way to connect Astro to an external data source is by using the data source's publicly accessible API endpoints.

> These endpoints can be configured by:

Author: @dnarain-astro Probably a question for you

> These endpoints can be configured by:

Reviewer: Yes, we should clarify by saying "The traffic flows directly between the data plane and the API endpoint and is not routed through the control plane"

- Setting [Environment Variables](environment-variables.md) with your connection information.
- Creating an [Airflow Connection](https://airflow.apache.org/docs/apache-airflow/stable/howto/connection.html).

### IP Whitelisting

Some data services, such as Snowflake and Databricks, provide an additional layer of security by requiring you to whitelist a specific IP address before that address can access the service.

> Each Cluster has a unique external IP address that will persist for the life of the Cluster. To retrieve the IP for a given Cluster, open a ticket with [Astronomer support](https://support.astronomer.io) and request it. If you have more than one Cluster, you will need to whitelist each Cluster individually on your data service provider.

Author: Are there functionality considerations given the two IPs? Do you only need to request one per allowlist? cc @dnarain-astro

> Each Cluster has a unique external IP address that will persist for the life of the Cluster. To retrieve the IP for a given Cluster, open a ticket with [Astronomer support](https://support.astronomer.io) and request it. If you have more than one Cluster, you will need to whitelist each Cluster individually on your data service provider.

Reviewer 2: No they'd need to whitelist both. Traffic may egress from one or the other depending which node the task is running on.

## VPC Peering

Each Cluster runs in a dedicated VPC. To set up private connectivity between the Astro VPC and a target VPC, you can set up a VPC peering connection. Peered VPCs provide additional security by ensuring private connectivity, reduced network transit costs, and simplified network layouts.

To create a VPC peering connection between a Cluster's VPC and a target VPC, open a support ticket with Astronomer and provide the following information for the target VPC:

- AWS account
- AWS Region

> - Name of the Cluster

Author: Which of these details is about your Astro Cluster, and which is about target VPC? @dnarain-astro

> - VPC CIDR

Author: cc @dnarain-astro

> - VPC CIDR

Reviewer: Its the "about your cluster" verbiage that makes this confusing. HEre's what I recommend:

To create a VPC peering connection between a Cluster's VPC and a target VPC, open a support ticket with Astronomer and provide the following information:
* Name of your Astro Cluster
* AWS Account for the target VPC
* AWS Region for the target VPC
* VPCID of for the target vpc
* CIDR of the target VPC

- Public Endpoints
- VPC Peering
- Transit Gateways
- Private Endpoints (AWS PrivateLink)

## Public Endpoints

The fastest and easiest way to connect Astro to an external data source is by using the data source's publicly accessible API endpoints.

These endpoints can be configured in either of the following ways:

- Set [Environment Variables](environment-variables.md) with your connection information.
- Create an [Airflow Connection](https://airflow.apache.org/docs/apache-airflow/stable/howto/connection.html).

### IP Whitelisting

Some data services, such as Snowflake and Databricks, provide an additional layer of security by requiring you to whitelist a specific IP address before that address can access the service.

Each Cluster has a unique external IP address that will persist for the life of the Cluster. To retrieve the IP for a given Cluster, open a ticket with Astronomer Support and request it. If you have more than one Cluster, you will need to whitelist each Cluster individually on your data service provider.

## VPC Peering

> Each Cluster runs in a dedicated VPC. To set up private connectivity between the Astro VPC and your VPC, you can set up a VPC peering connection. Peered VPCs provide additional security by ensuring private connectivity, reduced network transit costs, and simplified network layouts.

Author: ```suggestion
Each Cluster runs in a dedicated VPC. To set up private connectivity between the Astro VPC and a target VPC, you can set up a VPC peering connection. Peered VPCs provide additional security by ensuring private connectivity, reduced network transit costs, and simplified network layouts.
```

> Each Cluster runs in a dedicated VPC. To set up private connectivity between the Astro VPC and a target VPC, you can set up a VPC peering connection. Peered VPCs provide additional security by ensuring private connectivity, reduced network transit costs, and simplified network layouts.

Author: What is the typical scope/ context of a VPC connection? Will people be connecting to many VPCs, each its own service, a more general organization-level VPC, or both?

Reviewer: Sadly, the answer is both. Most people will have a VPC per service (e.g. a VPC for databricks, a VPC for redshift) and we expect customers to connect to multiple VPCs.

To create a VPC peering connection between a Cluster's VPC and a target VPC, open a support ticket with Astronomer and provide the following information for the target VPC:

- AWS account
- AWS Region
- Name of the Cluster
- VPC CIDR

In response, Astronomer will initiate a peering request. This peering request must be accepted by the owner of the target VPC.

Once peering is set up, continue to work with [Astronomer support](https://support.astronomer.io) and the owner of the target VPC to update the routing tables of both VPCs to direct traffic to each other.

> ### DNS Considerations with VPC Peering

Author: Is this topic something that all orgs will have to set up for their target VPC? Or is this just a specific way to connect?

Reviewer: I don't have a good answer to this one. My expectation is that large customers will use Private Hosted Zones and smaller ones are just using whatever defaults Amazon has in place.  But, we don't have enough usage data to prove or disprove my theory.

---
sidebar_label: "Connect to External Services"
title: "Connect Astro to External Data Services"
id: connectivity

> description: Learn how to connect Astro to different types of external data sources.

Author: This doc might encompass more types of connectivity in the future, I would keep this general.

These endpoints can be configured by:

- Setting [Environment Variables](environment-variables.md) with your endpoint information.
- Creating an [Airflow Connection](https://airflow.apache.org/docs/apache-airflow/stable/howto/connection.html).

Public connection traffic moves directly between your Data Plane and the external data source's API endpoint, meaning that the data in this traffic never reaches your Astronomer-managed Control Plane.

### IP Allowlist

Some data services, such as Snowflake and Databricks, provide an additional layer of security by requiring you to allowlist a specific IP address before you can access the service.

Each Cluster has a pair of unique external IP address that will persist for the life of the Cluster. These IP addresses are assigned for network address translation, meaning that they're responsible for all outbound traffic from your Cluster to the internet. To retrieve an IP for a given Cluster, open a ticket with [Astronomer support](https://support.astronomer.io) and request it. If you have more than one Cluster, you will need to allowlist each Cluster individually on your data service provider.

## VPC Peering

> Each Cluster runs in a dedicated VPC. To set up private connectivity between the Astro VPC and a target VPC, you can set up a VPC peering connection. Peered VPCs provide additional security by ensuring private connectivity, reduced network transit costs, and simplified network layouts.

Reviewer 3: ```suggestion
Each Astro Cluster runs in a dedicated VPC. To set up private connectivity between an Astro VPC and another VPC, you can set up a VPC peering connection. Peered VPCs provide additional security by ensuring private connectivity, reduced network transit costs, and simplified network layouts.
```

Author: If we capitalize "Cluster" I don't think we need to refer to it as "Astro Cluster" beyond the first reference. Added to style guide https://www.notion.so/astronomerio/Astronomer-Industry-Terms-Style-Guide-5d2389d462734e149d2e66a5cf606228

> To create a VPC peering connection between a Cluster's VPC and a target VPC, open a support ticket with Astronomer and provide the following information:

Reviewer 3: ```suggestion
To create a VPC peering connection between an Astro Cluster's VPC and a target VPC, reach out to [Astronomer support](https://support.astronomer.io) and provide the following information:
```

> - Name of your Cluster
> - AWS account for the target VPC
> - AWS Region for the target VPC
> - VPC ID of the target VPC
> - CIDR of the target VPC

Reviewer 3: ```suggestion
- Name of your Astro Cluster
- AWS Account ID for the target VPC
- AWS Region for the target VPC
- VPC ID of the target VPC
- CIDR of the target VPC
```

Assume Account ID here?

In response, Astronomer will initiate a peering request. To connect succesfully, this peering request must be accepted by the owner of the target VPC.

> Once peering is set up, continue to work with [Astronomer support](https://support.astronomer.io) and the owner of the target VPC to update the routing tables of both VPCs to direct traffic to each other.

Reviewer 3: I don't feel like we need to link to support here since we already did and this sentence isn't specifically actionable
